app/assets/javascripts/discourse/app/components/d-multi-select.gjs

```diff
@@ -0,0 +1,256 @@
+  }
+
+  @action
+  toggle(result, event) {
+    event?.stopPropagation();
+
+    if (this.isSelected(result)) {
+      this.selection = new TrackedArray(
+        this.selection.filter((item) => !this.compare(item, result))
+      );
+    } else {
+      this.selection.push(result);
+    }
+
+    this.args.onChange?.(this.selection);
+  }
+
+  @action
+  compare(a, b) {
+    return this.args.compareFn?.(a, b) ?? a.id === b.id;
+  }
+
+  <template>
+    <DMenu
+      @identifier="d-multi-select"
+      @triggerComponent={{element "div"}}
+      @triggerClass={{concatClass (if this.hasSelection "--has-selection")}}
+      ...attributes
+    >
+      <:trigger>
+        {{#if this.selection}}
+          <div class="d-multi-select-trigger__selection">
+            {{#each this.selection as |item|}}
+              <button
+                class="d-multi-select-trigger__selected-item"
+                {{on "click" (fn this.remove item)}}
+              >
+                <span class="d-multi-select-trigger__selection-label">{{yield
+                    item
+                    to="selection"
+                  }}</span>
+                {{icon
+                  "xmark"
+                  class="d-multi-select-trigger__remove-selection-icon"
+                }}
+              </button>
+            {{/each}}
+          </div>
+        {{else}}
+          <span class="d-multi-select-trigger__label">{{this.label}}</span>
+        {{/if}}
+
+        <DButton
+          @icon="angle-down"
+          class="d-multi-select-trigger__expand-btn btn-transparent"
+          @action={{@componentArgs.show}}
+        />
+      </:trigger>
+      <:content>
+        <DropdownMenu class="d-multi-select__content" as |menu|>
+          <menu.item class="d-multi-select__search-container">
+            {{icon "magnifying-glass"}}
+            <TextField
+              class="d-multi-select__search-input"
+              autocomplete="off"
+              @placeholder={{i18n "multi_select.search"}}
+              @type="search"
+              {{on "input" this.search}}
+              {{on "keydown" this.handleKeydown}}
+              {{didInsert this.focus}}
+              @value={{readonly this.searchTerm}}
+            />
+          </menu.item>
+
+          <menu.divider />
+
+          <AsyncContent
+            @asyncData={{fn @loadFn this.searchTerm}}
+            @debounce={{true}}
+            @context={{this.searchTerm}}
+          >
+            <:empty><div class="d-multi-select__search-no-results">{{i18n
+                  "multi_select.no_results"
+                }}</div></:empty>
+            <:loading>
+              <div class="d-multi-select__skeletons">
+                <Skeleton />
+                <Skeleton />
+                <Skeleton />
+                <Skeleton />
+                <Skeleton />
+              </div>
+            </:loading>
+            <:content as |results|>
+              {{this.resultsChanged results}}
```

Comment:
This is essentially creating an observer... which isn't ideal. I think we should try to find a way to model this declaratively. Maybe it would help to use ember-async-data directly, instead of using our simplified `AsyncContent` wrapper?

Reply:
Yes I know it looks like an observer, and that was my initial idea to use ember-async-data directly, but left it here to see if you have a better idea or something with `AsyncContent` I don't know

Reply:
Yeah I think this use-case is more complex than AsyncContent is designed for, so using ember-async-data directly should allow for something cleaner 🤞